sdk/src/asset_handlers/mp3_io.rs

```diff
@@ -0,0 +1,480 @@
+
+        // copy temp file to asset
+        std::fs::rename(temp_file.path(), asset_path)
+            // if rename fails, try to copy in case we are on different volumes
+            .or_else(|_| std::fs::copy(temp_file.path(), asset_path).and(Ok(())))
+            .map_err(Error::IoError)
+    }
+
+    fn get_object_locations(
+        &self,
+        asset_path: &std::path::Path,
+    ) -> Result<Vec<HashObjectPositions>> {
+        let mut f = std::fs::File::open(asset_path).map_err(|_err| Error::EmbeddingError)?;
+
+        self.get_object_locations_from_stream(&mut f)
+    }
+
+    fn remove_cai_store(&self, asset_path: &Path) -> Result<()> {
+        self.save_cai_store(asset_path, &[])
+    }
+
+    fn remote_ref_writer_ref(&self) -> Option<&dyn RemoteRefEmbed> {
+        None
+    }
+
+    fn supported_types(&self) -> &[&str] {
+        &SUPPORTED_TYPES
+    }
+}
+
+impl CAIWriter for Mp3IO {
+    fn write_cai(
+        &self,
+        input_stream: &mut dyn CAIRead,
+        output_stream: &mut dyn CAIReadWrite,
+        store_bytes: &[u8],
+    ) -> Result<()> {
+        let header = ID3V2Header::read_header(input_stream)?;
+        input_stream.rewind()?;
+
+        let mut out_tag = Tag::new();
+
+        // wrapper to protect imput stream from being gobbled
+        let reader = CAIReadWrapper {
+            reader: input_stream,
+        };
+
+        if let Ok(tag) = Tag::read_from(reader) {
+            for f in tag.frames() {
+                match f.content() {
+                    // remove existing manifest keeping existing frames
+                    Content::EncapsulatedObject(eo) => {
+                        if eo.mime_type != "application/x-c2pa-manifest-store" {
+                            out_tag.add_frame(f.clone());
+                        }
+                    }
+                    _ => {
+                        out_tag.add_frame(f.clone());
+                    }
+                }
+            }
+        }
+
+        // only add new tags
+        if !store_bytes.is_empty() {
+            // Add new manifest store
+            let frame = Frame::with_content(
+                "GEOB",
+                Content::EncapsulatedObject(EncapsulatedObject {
+                    mime_type: GEOB_FRAME_MIME_TYPE.to_string(),
+                    filename: GEOB_FRAME_FILE_NAME.to_string(),
+                    description: GEOB_FRAME_DESCRIPTION.to_string(),
+                    data: store_bytes.to_vec(),
+                }),
+            );
+
+            out_tag.add_frame(frame);
+        }
+
+        // wrapper to protect output stream from being gobbled
+        let writer = CAIReadWriteWrapper {
+            reader_writer: output_stream,
+        };
+
+        // write new tag to output stream
+        out_tag
+            .write_to(writer, Version::Id3v24)
+            .map_err(|_e| Error::BadParam("could not write to output stream".to_string()))?;
```

Comment:
I don't think it's right to call this a `BadParam`. Is there something like an I/O error that we can pass through?

Reply:
Agreed.